### PR TITLE
Master wowl documents wbr

### PR DESCRIPTION
**Pull request by Williambraecky**
WIP

### CODE REVIEW

**Comment by Williambraecky** [comment]
Hello @alexkuhn 🙂 
The task is still in testing stage but it would be nice if you could take a look at those changes (like we discussed on discord recently)
Cheers

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml [hunk 1]
```diff
@@ -5,24 +5,25 @@
         <t t-if="attachmentViewer">
             <div class="o_AttachmentViewer flex-column align-items-center d-flex w-100 h-100" t-attf-class="{{ className }}" t-on-click="attachmentViewer.onClick" t-on-keydown="_onKeydown" tabindex="0" t-ref="root">
                 <div class="o_AttachmentViewer_header d-flex w-100 bg-black-75 text-400" t-on-click="attachmentViewer.onClickHeader">
-                    <t t-if="attachmentViewer.attachment.isViewable">
+                    <t t-set="attachment" t-value="attachmentViewer.attachment"/>
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml:8]: Don't add this `t-set`, this is introducing a lot of unnecessary diff.

File: addons/mail/static/src/models/attachment_viewer.js [hunk 2]
```diff
@@ -213,6 +232,16 @@ registerModel({
...
+        /**
+         * @private
+         * @returns {boolean}
+         */
+        _computeIsMultiAttachment() {
+            if (this.attachments) {
+                return this.attachments.length > 1;
+            }
+            return false;
+         }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer.js:244]: This field is kind of useless, why not check attachments length directly like before?

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml (hunk 1 above)
Comment (Williambraecky) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.xml:8]: In documents we may have documents that have no attachment (youtube links for example) as such it was necessary to avoid rewriting all lines that use `attachmentViewer.attachment`.
This is the same reasoning behind the isMultiAttachment field.
Another way would be to make a computed field such as `attachmentOrDocument` but it does not make much sense when documents is not installed.
What do you think?

File: addons/mail/static/src/models/attachment_viewer.js (hunk 2 above)
Comment (Williambraecky) [reply addons/mail/static/src/models/attachment_viewer.js:244]: Answered on other comment :)

File: addons/mail/static/src/models/attachment_viewer.js [hunk 3]
```diff
@@ -221,15 +240,14 @@ registerModel({
...
+        attachment: one('AttachmentViewerItem', {
+            compute: '_computeAttachment',
         }),
         attachmentList: one('AttachmentList', {
             related: 'dialogOwner.attachmentListOwnerAsAttachmentView',
-            required: true,
         }),
-        attachments: many('Attachment', {
-            related: 'attachmentList.viewableAttachments',
+        attachments: many('AttachmentViewerItem', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer.js:249]: `attachment` and `attachments` should be renamed according to the new model (`attachmentViewerItem`) to avoid confusion

File: addons/mail/static/src/models/attachment_viewer.js [hunk 4]
```diff
@@ -179,19 +180,37 @@ registerModel({
...
+                    return { attachmentOwner: replace(attachment) };
+                }));
+            }
+            return clear();
+        },
         /**
          * @private
          * @returns {string}
          */
         _computeImageUrl() {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer.js:209]: This compute is not very useful, I think we can remove the field and directly read from the item.

File: addons/mail/static/src/models/attachment_viewer_item.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
+/** @odoo-module **/
+
+import { registerModel } from "@mail/model/model_core";
+import { attr, one } from "@mail/model/model_field";
+
+registerModel({
+    name: "AttachmentViewerItem",
+    identifyingFields: [["attachmentOwner"]],
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer_item.js:8]: Maybe rename `AttachmentViewerViewable`.

I feel like the "item" concept should be strongly related to the parent model, whereas here it's just related to the attachment/document generally speaking.

Each attachment/document is a `AttachmentViewerViewable` regardless of it's actually being viewed at a point in time.

File: addons/mail/static/src/models/attachment_viewer.js [hunk 6]
```diff
@@ -221,15 +232,14 @@ registerModel({
         angle: attr({
             default: 0,
         }),
-        attachment: one('Attachment', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer.js:224]: other occurrences of removed fields need to be adapted similar to the template (example `onClickDownload` still uses `attachment`)

Comment (Williambraecky) [reply addons/mail/static/src/models/attachment_viewer.js:224]: Oups my bad

File: addons/mail/static/src/models/attachment_viewer_viewable.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+import { registerModel } from "@mail/model/model_core";
+import { attr, one } from "@mail/model/model_field";
+
+registerModel({
+    name: "AttachmentViewerViewable",
+    identifyingFields: [["attachmentOwner"]],
+    recordMethods: {
+        download() {
+            if (this.attachmentOwner) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer_viewable.js:11]: Aren't all the conditions on `attachmentOwner` unnecessary in this file?
The override is checking for its own type and then just calling super if needed, which would always fall back to having an attachment?

Comment (Williambraecky) [reply addons/mail/static/src/models/attachment_viewer_viewable.js:11]: True

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js [hunk 8]
```diff
@@ -93,13 +93,13 @@ export class AttachmentViewer extends Component {
      * @private
      */
     _handleImageLoad() {
-        if (!this.attachmentViewer.exists() || !this.attachmentViewer.attachment) {
+        if (!this.attachmentViewer.exists() || !this.attachmentViewer.attachmentViewerViewable) {
             return;
         }
         const refs = this._getRefs();
-        const image = refs[`image_${this.attachmentViewer.attachment.id}`];
+        const image = refs[`image_${this.attachmentViewer.attachmentViewerViewable.id}`];
```
Comment (seb-odoo) [inline addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:100]: I believe this is no longer correct, `id` of attachment could conflict with `id` of document but we would need separate `ref` in that case.

Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:100]: consider using `localId` instead of `id` for the `ref` instead

File: addons/mail/static/src/models/attachment_viewer.js [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 import { registerModel } from '@mail/model/model_core';
 import { attr, many, one } from '@mail/model/model_field';
+import { clear, insertAndReplace, replace } from '../model/model_field_command';
```
Comment (seb-odoo) [inline addons/mail/static/src/models/attachment_viewer.js:5]: prefer using full path, relative is a nightmare to maintain

File: addons/mail/static/src/components/attachment_viewer/attachment_viewer.js (hunk 8 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/attachment_viewer/attachment_viewer.js:100]: It's just a unique identifier we have for each record. It's currently based on model name / identifying fields, but this is a technical choice to help debug, but this could as well be a unique hash.

**Review by seb-odoo — APPROVED** [review]
Ok conceptually for the community part.

Name `AttachmentViewerViewable` is still up to change to but ok as a first version, let's not get stuck on it too long.

**Comment by tivisse** [comment]
@robodoo r+
